slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -276,6 +276,55 @@ ::grpc::Status SLAMServiceImpl::GetCurrentPointCloudMap(
     }
 }
 
+// GetPointCloudMap returns the current sampled pointcloud derived from the
+// painted map, using probability estimates
+::grpc::Status SLAMServiceImpl::GetPointCloudMapStream(
+    ServerContext *context, const GetPointCloudMapStreamRequest *request,
+    ServerWriter<GetPointCloudMapStreamResponse> *writer) {
+    return ::grpc::Status(::grpc::StatusCode::UNIMPLEMENTED, "");
+}
+
+// GetInternalState returns the current internal state of the map which is
+// a pbstream for cartographer. The internal state is streamed in chunks of size
+// maximumGRPCByteChunkSize
+::grpc::Status SLAMServiceImpl::GetInternalStateStream(
+    ServerContext *context, const GetInternalStateStreamRequest *request,
+    ServerWriter<GetInternalStateStreamResponse> *writer) {
+    boost::uuids::uuid uuid = boost::uuids::random_generator()();
+    std::string filename = path_to_map + "/" + "temp_internal_state_" +
+                           boost::uuids::to_string(uuid) + ".pbstream";
+    {
+        std::lock_guard<std::mutex> lk(map_builder_mutex);
+        bool ok = map_builder.SaveMapToFile(true, filename);
+        if (!ok) {
+            std::ostringstream oss;
+            oss << "Failed to save the state as a pbstream.";
+            return grpc::Status(grpc::StatusCode::UNAVAILABLE, oss.str());
+        }
+    }
+
+    std::string buf;
+    try {
+        ConvertSavedMapToStream(filename, &buf);
+    } catch (std::exception &e) {
+        std::ostringstream oss;
+        oss << "error during data serialization: " << e.what();
+        return grpc::Status(grpc::StatusCode::UNAVAILABLE, oss.str());
+    }
+
+    std::string internal_state_chunk;
+    GetInternalStateStreamResponse response;
+    for (int start_index = 0; start_index < buf.size();
+         start_index += maximumGRPCByteChunkSize) {
+        internal_state_chunk =
+            buf.substr(start_index, maximumGRPCByteChunkSize);
```

Comment:
Will this be out of range for the last chunk (which might have size less than maximumGRPCByteChunkSize)?

Reply:
if the chunk is smaller than `maximumGRPCByteChunkSize`, `substr` will use the remainder of the characters.